core/trino-spi/src/main/java/io/trino/spi/connector/SourcePage.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi.connector;
+
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+
+import java.util.function.ObjLongConsumer;
+
+/**
+ * A page of data from a connector.
+ * <p>
+ * A page has a fixed number of positions and a fixed set of channels.
+ * <p>
+ * This interface is not thread safe.
+ */
+public interface SourcePage
+{
+    /**
+     * Creates a new SourcePage from the specified block.
+     */
+    static SourcePage create(int positionCount)
+    {
+        return new NoChannelsSourcePage(positionCount);
+    }
+
+    /**
+     * Creates a new SourcePage from the specified block.
+     */
+    static SourcePage create(Block block)
+    {
+        return new StaticSourcePage(new Page(block.getPositionCount(), block));
+    }
+
+    /**
+     * Creates a new SourcePage from the specified page.
+     */
+    static SourcePage create(Page page)
+    {
+        return new StaticSourcePage(page);
+    }
+
+    /**
+     * Gets the number of positions in the page.
+     */
+    int getPositionCount();
```

Comment:
With `SourcePage` now being mutable, there's a potential issue where you might retrieve the `positionCount`, but then another operation (like calling `selectPositions`) alters the source, causing the positions to no longer align with the current state.

Reply:
Yes.  There are lots of scenarios where you can get in trouble.  The interface is single threaded so there should be no worries about external actors modifying the contents.  The interface design is a compromise between simple usability and performance.  
I considered designs where you select positions resulted in a new object, but it has the problem that it does not allow the reading code to skip data, because the original object still exists and someone may decide to use that object. 
Users of this interface need to be aware of what they are doing, and if they don't want to deal with stuff chaning they can simply materialize the whole page.

Reply:
On the other hand, can we leverage this behavior and avoid setting the position count until we materialize the Page?
e.g.
When a page needs to return as many rows as possible while keeping the total size under 1MB, determining the number of positions is straightforward if the page contains only fixed-size columns. However, if it includes non-fixed-size columns, the number of rows must be estimated, typically using a worst-case scenario.
if we won't need to commit the positionCount for the SourcePage this problem can be solved

Reply:
Maybe.  The implementation could delay the position count determinization until this method is called... but the position count is needed when any block is fetched, so I'm not sure if this will help as much as you thing.  The most common scenario will be:
1. execute filter - load one or more blocks and filter
2. select filtered positions - reduce page to a set of positions
3. project remaining blocks - load the remaining blocks for the selected positions
or there is no filtering so all blocks just get loaded.  Either way, the first piece of information you need is the number of positions to return.